testing/middleware/engine-api-proxy/proxy.go

```diff
@@ -0,0 +1,237 @@
+func New(opts ...Option) (*Proxy, error) {
+	p := &Proxy{
+		cfg: &config{
+			proxyHost: defaultProxyHost,
+			proxyPort: defaultProxyPort,
+			logger:    logrus.New(),
+		},
+		interceptors: make(map[string]*interceptorConfig),
+	}
+	for _, o := range opts {
+		if err := o(p); err != nil {
+			return nil, err
+		}
+	}
+	if p.cfg.destinationUrl == nil {
+		return nil, errors.New("must provide a destination address for request proxying")
+	}
+	mux := http.NewServeMux()
+	mux.Handle("/", p)
+	addr := fmt.Sprintf("%s:%d", p.cfg.proxyHost, p.cfg.proxyPort)
+	srv := &http.Server{
+		Handler: mux,
+		Addr:    addr,
+	}
+	p.address = addr
+	p.srv = srv
+	return p, nil
+}
+
+// Address for the proxy server.
+func (p *Proxy) Address() string {
+	return p.address
+}
+
+// Start a proxy server.
+func (p *Proxy) Start(ctx context.Context) error {
+	p.srv.BaseContext = func(listener net.Listener) context.Context {
+		return ctx
+	}
+	p.cfg.logger.WithFields(logrus.Fields{
+		"forwardingAddress": p.cfg.destinationUrl.String(),
+	}).Infof("Engine proxy now listening on address %s", p.address)
+	go func() {
+		if err := p.srv.ListenAndServe(); err != nil {
+			p.cfg.logger.Error(err)
+		}
+	}()
+	for {
+		<-ctx.Done()
+		return p.srv.Shutdown(context.Background())
+	}
+}
+
+// ServeHTTP requests from a consensus client to an execution client, modifying in-flight requests
+// and/or responses as desired. It also processes any backed-up requests.
+func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	requestBytes, err := parseRequestBytes(r)
+	if err != nil {
+		p.cfg.logger.WithError(err).Error("Could not parse request")
+		return
+	}
+	// Check if we need to intercept the request with a custom response.
+	hasIntercepted, err := p.interceptIfNeeded(requestBytes, w)
+	if err != nil {
+		p.cfg.logger.WithError(err).Error("Could not intercept request")
+		return
+	}
+	if hasIntercepted {
+		return
+	}
+	// If we are not intercepting the request, we proxy as normal.
+	p.proxyRequest(requestBytes, w, r)
+}
+
+// AddRequestInterceptor for a desired json-rpc method by specifying a custom response
+// and a function that checks if the interceptor should be triggered.
+func (p *Proxy) AddRequestInterceptor(rpcMethodName string, response interface{}, trigger func() bool) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	p.interceptors[rpcMethodName] = &interceptorConfig{
+		response,
+		trigger,
+	}
+}
+
+// Checks if there is a custom interceptor hook on the request, check if it can be
+// triggered, and then write the custom response to the writer.
+func (p *Proxy) interceptIfNeeded(requestBytes []byte, w http.ResponseWriter) (hasIntercepted bool, err error) {
+	if !isEngineAPICall(requestBytes) {
+		return
+	}
+	var jreq *jsonRPCObject
+	jreq, err = unmarshalRPCObject(requestBytes)
+	if err != nil {
+		return
+	}
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	interceptor, shouldIntercept := p.interceptors[jreq.Method]
```

Comment:
What do you think about having the ability for a method type to have multiple interceptors ? That way we could run more complex setups if necessary.

Reply:
Good idea, but I think I want to keep this PR smaller for now and we can add more functionality as we need